Purchases/Public/RCPurchases.m

```diff
@@ -978,17 +989,20 @@ - (void)handleOfferingsUpdateError:(NSError *)error completion:(RCReceiveOfferin
 }
 
 - (void)receiptData:(RCReceiveReceiptDataBlock)completion {
-    [self receiptDataWithForceRefresh:NO completion:completion];
+    [self receiptDataWithReceiptRefreshPolicy:RCReceiptRefreshPolicyOnlyIfEmpty
+                                   completion:completion];
 }
 
-- (void)receiptDataWithForceRefresh:(BOOL)forceRefresh completion:(RCReceiveReceiptDataBlock)completion {
-    if (forceRefresh) {
+- (void)receiptDataWithReceiptRefreshPolicy:(RCReceiptRefreshPolicy)refreshPolicy
+                                 completion:(RCReceiveReceiptDataBlock)completion {
+    if (refreshPolicy == RCReceiptRefreshPolicyAlways) {
         RCDebugLog(@"Forced receipt refresh");
         [self refreshReceipt:completion];
         return;
     }
     NSData *receiptData = [self.receiptFetcher receiptData];
-    if (receiptData == nil || receiptData.length == 0) {
+    BOOL receiptIsEmpty = receiptData == nil || receiptData.length == 0;
+    if (receiptIsEmpty && refreshPolicy == RCReceiptRefreshPolicyOnlyIfEmpty) {
```

Comment:
I just realized that in `paymentDiscountForProductDiscount` we are not checking if the receipt is empty, we are just doing:

```
 [self receiptData:^(NSData *data) {
        [self.backend postOfferForSigning:discount.identifier
```

So this change would possibly return an empty receipt to that function, and we should probably return `missingReceiptFileError`

Reply:
@vegaro I'm confused, why wouldn't this change should keep the same behavior for `paymentDiscountForProductDiscount`?
It calls `receiptData`, which in turn calls `receiptDataWithReceiptRefreshPolicy:RCReceiptRefreshPolicyOnlyIfEmpty`, so this line would render the exact same result as before, right?

Reply:
talked about it over slack. 
Since within `paymentDiscountForProductDiscount` we're always calling with `RCReceiptRefreshPolicyOnlyIfEmpty`, there's no change in behavior. 

However, Cesar did correctly spot that there's an incorrect assumption in the method where we assume that the receiptData will be non-null and non-empty. 
if the receipt data is nil, that would cause a crash: we'd try to parse nil into a base-64 string, which would no-op and return nil again, and that nil would be inserted into a dictionary, causing a crash because NSDictionaries can't have nil values. 

if it's empty, it would actually post to the backend, but with an empty string, which is an unnecessary backend call. And every time you call the backend without needing to, a panda dies. Don't kill pandas. 

I'm fixing this and adding a test case for it.

Reply:
updated: https://github.com/RevenueCat/purchases-ios/pull/387/files/4a4186700f03dcb401e6fb8842fed2592472c872..ac8575073a1f87135b69a283896d0889b6f750e3